### PR TITLE
Implement ATA and port python-pure25519's curve check

### PR DESCRIPTION
**Pull request by liyanhrxy**

### CODE REVIEW

File: solana/utils/ed25519_base.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
+Q = 2 ** 255 - 19
```
Comment (michaelhly) [inline solana/utils/ed25519_base.py:1]: Could you site where you got this file from? i.e. Add link to original source.

**Review by michaelhly — COMMENTED** [review]
Please run `make format`.

**Comment by liyanhrxy** [comment]
> Please run `make format`.

OK

File: solana/utils/ed25519_base.py (hunk 1 above)
Comment (liyanhrxy) [reply solana/utils/ed25519_base.py:1]: https://github.com/warner/python-pure25519/blob/master/pure25519/basic.py#L132

File: solana/utils/ed25519_base.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
```
Comment (michaelhly) [inline solana/utils/ed25519_base.py:1]:
```suggestion
"""curve25519/ed25519 helpers.

Sourced from https://github.com/warner/python-pure25519/blob/master/pure25519/basic.py
"""
Q = 2 ** 255 - 19
```

File: solana/utils/ed25519_base.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+L = 2 ** 252 + 27742317777372353535851937790883648493
+
+
+def inv(x):
```
Comment (michaelhly) [inline solana/utils/ed25519_base.py:5]:
```suggestion
def inv(x):  # noqa: D103
```

File: solana/utils/ed25519_base.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    return pow(x, Q - 2, Q)
+
+
+d = -121665 * inv(121666)
+I = pow(2, (Q - 1) // 4, Q)
+
+
+def xrecover(y):
```
Comment (michaelhly) [inline solana/utils/ed25519_base.py:13]:
```suggestion
def xrecover(y): # noqa: D103
```

File: solana/utils/ed25519_base.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    xx = (y * y - 1) * inv(d * y * y + 1)
+    x = pow(xx, (Q + 3) // 8, Q)
+    if (x * x - xx) % Q != 0:
+        x = (x * I) % Q
+    if x % 2 != 0:
+        x = Q - x
+    return x
+
+
+def _isoncurve(P):
```
Comment (michaelhly) [inline solana/utils/ed25519_base.py:23]:
```suggestion
def _isoncurve(P):  # noqa: D103
```

File: solana/utils/ed25519_base.py [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+    x = P[0]
+    y = P[1]
+    return (-x * x + y * y - 1 - d * x * x * y * y) % Q == 0
+
+
+class NotOnCurve(Exception):
```
Comment (michaelhly) [inline solana/utils/ed25519_base.py:29]:
```suggestion
class NotOnCurve(Exception):  # noqa: D101
```

File: solana/utils/ed25519_base.py [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+    x = xrecover(y)
+    if bool(x & 1) != bool(unclamped & (1 << 255)):
+        x = Q - x
+    P = [x, y]
+    if not _isoncurve(P):
+        raise NotOnCurve("decoding point that is not on curve")
+    return P
+
+
+def is_on_curve(s: bytes) -> bool:
```
Comment (michaelhly) [inline solana/utils/ed25519_base.py:45]:
```suggestion
def is_on_curve(s: bytes) -> bool:  # noqa: D103
```

**Review by michaelhly — COMMENTED** [review]
#noqa on pydocstyle linting

File: spl/token/instructions.py [hunk 8]
```diff
@@ -288,6 +290,17 @@ class Burn2Params(NamedTuple):
...
+    """Create associated token account transaction params."""
+
+    owner: PublicKey
+    """Owner of the associated token account."""
+    mint: PublicKey
+    """Public key of the minter account."""
+    payer: PublicKey
+    """Account paying for the transaction."""
+
+
```
[the remark this answers is not in the document]
Comment (liyanhrxy) [reply spl/token/instructions.py:303]:
```python3
def test_find_program_address():
    """Test create associated_token_address."""
    program_id = PublicKey('BPFLoader1111111111111111111111111111111111')
    program_address, nonce = PublicKey.find_program_address([bytes()], program_id)
    assert program_address == PublicKey.create_program_address([bytes(), helpers.to_uint8_bytes(nonce)], program_id)

```
what you want like this ??

[the remark this answers is not in the document]
Comment (michaelhly) [reply spl/token/instructions.py:303]: yes. Exactly.

[the remark this answers is not in the document]
Comment (liyanhrxy) [reply spl/token/instructions.py]: @michaelhly   all done, review please

**Review by michaelhly — APPROVED** [review]
Thank you both @liyanhrxy and @DaveWK!

**Pull request merged**